sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+  kEpmpPermissionsLockedMachineReadUserRead,               /**< LRWX = 0b1100 */
+  kEpmpPermissionsLockedMachineReadExecuteUserReadExecute, /**< LRWX = 0b1101 */
+  kEpmpPermissionsLockedMachineReadWriteUserReadWrite,     /**< LRWX = 0b1110 */
+  kEpmpPermissionsLockedMachineAllUserAll,                 /**< LRWX = 0b1111 */
+} epmp_permissions_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
+ * addresses).
+ *
+ * The intention is that this data structure is used to disambiguate the
+ * addresses of regions. It is therefore recommended that `start` and `end` are
+ * always used as labels when declaring a region.
+ */
+typedef struct epmp_region {
+  uintptr_t start;
+  uintptr_t end;
+} epmp_region_t;
+
+/**
+ * EPMP entry index.
+ */
+typedef enum epmp_entry {
+  kEpmpEntry0,
+  kEpmpEntry1,
+  kEpmpEntry2,
+  kEpmpEntry3,
+  kEpmpEntry4,
+  kEpmpEntry5,
+  kEpmpEntry6,
+  kEpmpEntry7,
+  kEpmpEntry8,
+  kEpmpEntry9,
+  kEpmpEntry10,
+  kEpmpEntry11,
+  kEpmpEntry12,
+  kEpmpEntry13,
+  kEpmpEntry14,
+  kEpmpEntry15,
+} epmp_entry_t;
+
+/**
+ * EPMP entry count.
+ */
+typedef enum epmp_entry_count {
+  kEpmpEntryCount1,
+  kEpmpEntryCount2,
+  kEpmpEntryCount3,
+  kEpmpEntryCount4,
+  kEpmpEntryCount5,
+  kEpmpEntryCount6,
+  kEpmpEntryCount7,
+  kEpmpEntryCount8,
+  kEpmpEntryCount9,
+  kEpmpEntryCount10,
+  kEpmpEntryCount11,
+  kEpmpEntryCount12,
+  kEpmpEntryCount13,
+  kEpmpEntryCount14,
+  kEpmpEntryCount15,
+  kEpmpEntryCount16,
+} epmp_entry_count_t;
+
+/**
+ * EPMP transaction.
+ *
+ * Must be initialized using epmp_transaction_init before use.
+ *
+ * This structure should be considered opaque.
+ */
+typedef struct epmp_transaction {
+  epmp_entry_count_t remaining;
+  epmp_entry_count_t done;
+  epmp_entry_count_t total;  // remaining + done
+  // TODO: add checksum tracker
+} epmp_transaction_t;
+
+/**
+ * EPMP generic status codes.
+ *
+ * These error codes can be used by any function. However if a function
+ * requires additional status codes, it must define a set of status codes to
+ * be used exclusively by such function.
+ */
+typedef enum epmp_result {
+  kEpmpOk = 0, /**< Success. */
+  kEpmpError,  /**< General error. */
+  kEpmpBadArg, /**< Input parameter is not valid. */
+} epmp_result_t;
+
+/**
+ * Start an EPMP transaction.
+ *
+ * The transaction will be initialized to expect the given number of entries
+ * to be reconfigured. Only one transaction may be in progress at a time.
+ * Call epmp_transaction_push to finalize the transaction and update the
+ * EPMP control registers.
+ *
+ * @param transaction Transaction to initialize.
+ * @param count Number of regions to be configured.
+ * @returns `epmp_result_t` // TODO: more specialized return code?
+ */
+epmp_result_t epmp_transaction_init(epmp_transaction_t *transaction,
```

Comment:
Nit: this isn't a DIF, but normally we call transactional APIs something like `<module>_start` and `<module>_end`.